vnext/JSI/Shared/ChakraObjectRef.cpp

```diff
@@ -0,0 +1,325 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "ChakraObjectRef.h"
+
+#include "Unicode.h"
+#include "Utilities.h"
+
+#include "jsi//jsi.h"
+
+#include <memory>
+#include <sstream>
+#include <utility>
+
+namespace Microsoft::JSI {
+
+namespace {
+
+inline void CheckedJsAddRef(JsRef ref) {
+  ThrowUponChakraError(JsAddRef(ref, nullptr), "JsAddRef");
+}
+
+inline void CheckedJsRelease(JsRef ref) {
+  ThrowUponChakraError(JsRelease(ref, nullptr), "JsRelease");
+}
+
+} // namespace
+
+void ThrowUponChakraError(JsErrorCode error, const char *const chakraApiName) {
+  if (error != JsNoError) {
+    std::ostringstream errorString;
+    errorString << "A call to "
+                << (chakraApiName ? chakraApiName : "Chakra(Core) API")
+                << " returned error code 0x" << std::hex << error << '.';
+    throw facebook::jsi::JSINativeException(errorString.str());
+  }
+}
+
+ChakraObjectRef::ChakraObjectRef(const ChakraObjectRef &original) noexcept
+    : m_ref{original.m_ref}, m_state{original.m_state} {
+  if (m_state == State::Initialized) {
+    assert(m_ref);
```

Comment:
>assert [](start = 4, length = 6)

This is also related to error handling story. 

In general, Asserts are pretty useless, we should use ShipAssert or VerifyElseDump to expose errors to devs. 

Can you remove this assert?

I've created a task item about error handling:
[https://office.visualstudio.com/OC/_workitems/edit/3712909/](https://office.visualstudio.com/OC/_workitems/edit/3712909/)

In there I tried to summarize work that needs to be done. Could you take a quick look to see if it's accurate or if there is anything else you'd want to add?

Reply:
I took a look and added a comment on the work item.

---
In reply to: [336169600](https://github.com/microsoft/react-native-windows/pull/3445#discussion_r336169600) [](ancestors = 336169600)